homeassistant/components/sensor/apcupsd.py

```diff
@@ -0,0 +1,107 @@
+"""
+homeassistant.components.sensor.apcupsd
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Provides a sensor to track various status aspects of a UPS.
+"""
+import logging
+
+from homeassistant.core import JobPriority
+from homeassistant.const import TEMP_CELCIUS
+from homeassistant.helpers.entity import Entity
+from homeassistant.components import apcupsd
+
+
+DEPENDENCIES = [apcupsd.DOMAIN]
+
+DEFAULT_NAME = "UPS Status"
+
+SPECIFIC_UNITS = {
+    "ITEMP": TEMP_CELCIUS
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """
+    Ensure that the 'type' config value has been set and use a specific unit
+    of measurement if required.
+    """
+    typ = config.get(apcupsd.CONF_TYPE)
+    if typ is None:
+        _LOGGER.error(
+            "You must include a '%s' when configuring an APCUPSd sensor.",
+            apcupsd.CONF_TYPE)
+        return
+    typ = typ.upper()
+
+    # Get a status reading from APCUPSd and check whether the user provided
+    # 'type' is present in the output. If we're not able to check, then assume
+    # the user knows what they're doing.
+    # pylint: disable=broad-except
+    status = None
+    try:
+        status = apcupsd.GET_STATUS()
+        if typ not in status:
+            _LOGGER.error(
+                "Specified '%s' of '%s' does not appear in the APCUPSd status "
+                "output.", apcupsd.CONF_TYPE, typ)
+            return
+    except Exception as exc:
+        _LOGGER.warning(
+            "Unable to fetch initial value from ACPUPSd to check that '%s' is "
+            "a supported '%s': %s", typ, apcupsd.CONF_TYPE, exc)
+    unit = SPECIFIC_UNITS.get(typ)
+    add_entities((
+        Sensor(hass, config, unit=unit, initial_status=status),
+    ))
+
+
+def infer_unit(value):
+    """
+    If the value ends with any of the units from ALL_UNITS, split the unit
+    off the end of the value and return the value, unit tuple pair. Else return
+    the original value and None as the unit.
+    """
+    from apcaccess.status import ALL_UNITS
+    for unit in ALL_UNITS:
```

Comment:
I don't think I understand what you're suggesting. Wouldn't this just require us to iterate `ALL_UNITS` twice?

At the moment, if any of the values end with `unit` then at worst we'll have to iterate until the last `ALL_UNITS` element, at best we get it on the first iteration and return. If none of the values end with `unit` then we only have to iterate through `ALL_UNITS` once anyway, checking whether `value.endswith(unit)`.